### PR TITLE
Docker health

### PR DESCRIPTION
**Pull request by tjamet**
#### Short description of what this resolves:

Add support for docker based health checks

#### Changes proposed in this pull request:

This commit adds support for docker-managed health checks.
    
There are 2 ways of using it:
- either let docker manage the health check on its own and inspect the
   status thanks to the docker API
- or use the docker API to run a custom health check command in the
   container

### CODE REVIEW

**Comment by dipak-pawar** [comment]
@tjamet [`DockerHealthAwaitStrategyTest` ](https://github.com/arquillian/arquillian-cube/pull/1004/files#diff-a161aeba78b437d8d0d71c4c15abcb3f) looks flaky. You can see it's failing for some time on CI. Build [Log](https://travis-ci.org/arquillian/arquillian-cube/jobs/347198942#L1784). Can you please look into it?

File: docker/docker/src/test/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategyTest.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        } catch (Throwable t) {
+            System.out.printf("Failed to clean test container %s: %s\n", containerId, t.toString());
+        }
+    }
+
+    private void testHealthCheckResult(String path, boolean status) {
+        testAwait(path, status, null);
+    }
+
+    private void testAwait(String imageId, boolean status, String[] command) {
```
[the remark this answers is not in the document]
Comment (tjamet) [reply docker/docker/src/test/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategyTest.java]: done

File: docker/docker/src/test/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategyTest.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+            if (containerId != null) {
+                dockerRm(containerId);
+            }
+        }
+        dockerRmi(healthSuccessImageId);
+        dockerRmi(healthFailureImageId);
+    }
+
+    @Test
+    public void TestHealthCheckSuccess() {
```
[the remark this answers is not in the document]
Comment (tjamet) [reply docker/docker/src/test/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategyTest.java]: done

File: docker/docker/src/test/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategyTest.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+    private static List<String> containerIds;
+
+    @BeforeClass
+    public static void createDockerClient() {
+        String host = System.getenv("DOCKER_HOST");
+        if (host == null || host.equals("")) {
+            host = "unix:///var/run/docker.sock";
+        }
+        dockerClient = DockerClientBuilder
```
[the remark this answers is not in the document]
Comment (tjamet) [reply docker/docker/src/test/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategyTest.java]: I have used the default client constructor instead, watching the travis logs to see if that works in CI

**Comment by tjamet** [comment]
@dipak-pawar Thank you for your feedback, I have addressed your points and added documentation, but went for letting the docker client to handle the socket path configuration instead

File: docker/docker/src/main/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategy.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    private Cube<?> cube;
+    private DockerClient client;
+    private String[] command;
+
+    public DockerHealthAwaitStrategy(Cube<?> cube, DockerClientExecutor dockerClientExecutor, Await params) {
+        super(params.getSleepPollingTime());
+        this.cube = cube;
+        this.client = dockerClientExecutor.getDockerClient();
+
+        if (params.getIterations() != null) {
```
Comment (dipak-pawar) [inline docker/docker/src/main/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategy.java]: How about moving this logic inside contrustor to one method and using it.

```java
    private int getIterations(Await params) {
        if (params.getIterations() != null) {
            return params.getIterations();
        } else {
            return DEFAULT_POLL_ITERATIONS;
        }
    }
```

File: docker/docker/src/test/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategyTest.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class DockerHealthAwaitStrategyTest {
+
+    private static final String DOCKER_HOST = "DOCKER_HOST";
+
+    public static DockerClient dockerClient;
+    public static String healthSuccessImageId;
+    public static String healthFailureImageId;
```
Comment (dipak-pawar) [inline docker/docker/src/test/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategyTest.java]: can't we make all(`dockerClient`, `healthSuccessImageId`, `healthFailureImageId` variables private here?

File: docker/docker/src/test/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategyTest.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    @Before
+    public void setup() {
+        containerIds = new ArrayList<>();
+    }
+
+    @After
+    public void cleanContainers() {
+        for (int i = 0; i < containerIds.size(); i++) {
```
[the remark this answers is not in the document]
Comment (tjamet) [reply docker/docker/src/test/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategyTest.java]: done

File: docker/docker/src/test/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategyTest.java (hunk 5 above)
Comment (tjamet) [reply docker/docker/src/test/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategyTest.java]: done

File: docker/docker/src/main/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategy.java (hunk 4 above)
Comment (dipak-pawar) [reply docker/docker/src/main/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategy.java]: Can you check this last comment?

Comment (tjamet) [reply docker/docker/src/main/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategy.java]: sorry missed it
done

File: docker/docker/src/main/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategy.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+public class DockerHealthAwaitStrategy extends SleepingAwaitStrategyBase {
+
+    protected static final Logger log = Logger.getLogger(DockerHealthAwaitStrategy.class.getName());
+
+    public static final String TAG = "docker_health";
+
+    private static final int DEFAULT_POLL_ITERATIONS = 10;
+
+    private int pollIterations = DEFAULT_POLL_ITERATIONS;
```
Comment (dipak-pawar) [inline docker/docker/src/main/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategy.java]: This is not required as we are initializing it in constructor.

File: docker/docker/src/main/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategy.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+import java.util.logging.Logger;
+import org.arquillian.cube.docker.impl.client.config.Await;
+import org.arquillian.cube.docker.impl.docker.DockerClientExecutor;
+import org.arquillian.cube.docker.impl.util.Ping;
+import org.arquillian.cube.docker.impl.util.PingCommand;
+import org.arquillian.cube.spi.Cube;
+
+public class DockerHealthAwaitStrategy extends SleepingAwaitStrategyBase {
+
+    protected static final Logger log = Logger.getLogger(DockerHealthAwaitStrategy.class.getName());
```
Comment (dipak-pawar) [inline docker/docker/src/main/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategy.java]: can't be private?

File: docker/docker/src/main/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategy.java (hunk 7 above)
Comment (tjamet) [reply docker/docker/src/main/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategy.java]: done

File: docker/docker/src/main/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategy.java (hunk 8 above)
Comment (tjamet) [reply docker/docker/src/main/java/org/arquillian/cube/docker/impl/await/DockerHealthAwaitStrategy.java]: done

**Comment by rpau** [comment]
@dipak-pawar when are you planning to release arquillian with this await strategy?

**Comment by dipak-pawar** [comment]
@rpau We are planning to release new version of cube this week.

**Pull request merged**